### PR TITLE
Benchmark improvements

### PR DESCRIPTION
**Pull request by LeoGrin**
- Random search
- Simple tests for the monitor function
- joblib support (I think this won't work for memory monitoring, but should work for time monitoring. Would love feedback on this).

@LilianBoulard this might be useful for the gapencoder benchmark
@jovan-stojanovic

### CODE REVIEW

File: benchmarks/utils/monitor.py [hunk 1]
```diff
@@ -184,13 +199,18 @@ def exec_func(*args, **kwargs) -> Dict[str, List[float]]:
...
-                parametrization = (parametrize, ())
+                parametrization = (
+                    parametrize
+                    if n_random_search is None
+                    else random.sample(parametrize, n_random_search),
+                    (),
+                )
             else:
                 parametrization = list(product(parametrize))
+                if not (n_random_search is None):
```
Comment (LilianBoulard) [inline benchmarks/utils/monitor.py:210]: Can be simplified:
```suggestion
                if n_random_search is not None:
```

**Review by LilianBoulard — COMMENTED** [review]
Thanks for the PR :)

On the parallelism of runs, I originally set them to run sequentially so the time and memory monitoring would be accurate.
If there is a reliable way to measure those while running the instances in parallel, I'm all for it! Though I don't think the current implementation is. I changed the inner structure quite a bit in #593, it might be easier to integrate parallelism with these.

I like the random search idea, though we should also have a way to set the seed, for reproducibility.

**Comment by GaelVaroquaux** [comment]
Can you:
- merge with master
- add a changelog entry

File: benchmarks/tests/test_monitor.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
+import pytest
```
Comment (GaelVaroquaux) [inline benchmarks/tests/test_monitor.py:1]: Please add a short module-level docstring describing what this does

File: benchmarks/tests/test_monitor.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from benchmarks.utils import monitor, repr_func  # replace with the actual import path
```
Comment (GaelVaroquaux) [inline benchmarks/tests/test_monitor.py:2]: Out of curiosity: is this supposed to work without installation? I'm a bit surprised that it does.

File: benchmarks/utils/monitor.py [hunk 4]
```diff
@@ -1,11 +1,13 @@
 import tracemalloc
+import random
```
Comment (GaelVaroquaux) [inline benchmarks/utils/monitor.py:2]: Can we have a small module-level docstring that describe what this file does?

**Review by GaelVaroquaux — CHANGES_REQUESTED** [review]
A few comments. Happy to discuss IRL if useful

File: benchmarks/tests/test_monitor.py (hunk 3 above)
Comment (LilianBoulard) [reply benchmarks/tests/test_monitor.py:2]: Yes, it does work, since there is an `__init__` file in `benchmarks/` :)

**Pull request closed without merging**